Add buffering to DPRINT

sagarwalTT opened the pull request:
#14827 
#5945
#6859

This PR implements host-side and device-side buffering for DPRINT so that DPRINTs are not interleaved.

### Checklist
- [ ] Post commit CI passes
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

sagarwalTT committed:
Added test

sagarwalTT committed:
Implemented flushing on ENDL

sagarwalTT committed:
Added initial implementation of device-side buffering

sagarwalTT committed:
Cleanup

sagarwalTT committed:
Updated test

sagarwalTT committed:
Added functionality to flush remaining data when device is detached

sagarwalTT committed:
More work

sagarwalTT committed:
Cleanup

tt-dma reviewed (approved):
Looks good to me, can we just add a note in the docs to show that the buffers are flushed on newline and device close? Just so users aren't surprised

sagarwalTT commented on tt_metal/impl/debug/dprint_server.cpp:
This condition should only be true if the buffer doesn't end with a null terminating char. I don't think there should be a case where the device adds data that is not null terminated to the buffer, but I left this condition in just as a safety check.

sagarwalTT committed:
Updated doc

sagarwalTT commented:
> Looks good to me, can we just add a note in the docs to show that the buffers are flushed on newline and device close? Just so users aren't surprised

@tt-dma I've updated the doc. Can you take a look at it to see if I added it in the right place and whether I should add any more detail before I merge it in?

sagarwalTT committed:
Updated doc

nhuang-tt commented on tt_metal/impl/debug/dprint_server.cpp:
```diff
@@ -600,6 +664,14 @@ void DebugPrintServerContext::AttachDevice(Device* device) {
         }
     }
 
+    device_intermediate_streams_force_flush_lock_.lock();
```
@sagarwalTT I started seeing seeing an assertion on main with a debug build coming from here

```
export TT_METAL_DPRINT_CORES=all
export TT_METAL_DPRINT_ETH_CORES=all
export TT_METAL_DPRINT_CHIPS=0,1
``

sagarwalTT replied:
what is the assertion error you're encountering?

nhuang-tt replied:
Nvm. Something is wrong with the state of my system...
